File: SecureExam-SIST/compilerserver.py
import subprocess
import os
import shutil
from flask import Flask, request, jsonify
from flask_cors import CORS

app = Flask(__name__)
CORS(app)

# Directory to store temporary code files
CODE_DIR = "/tmp/code_exec"

# Ensure the directory exists
os.makedirs(CODE_DIR, exist_ok=True)

# Cleanup function to remove temporary files
def cleanup():
    if os.path.exists(CODE_DIR):
        shutil.rmtree(CODE_DIR)
        os.makedirs(CODE_DIR, exist_ok=True)

# Function to run Python code
def run_python(code):
    try:
        result = subprocess.run(
            ['python3', '-c', code],
            capture_output=True,
            text=True,
            timeout=5
        )
        return result.stdout or result.stderr
    except subprocess.TimeoutExpired:
        return "Error: Code execution timed out."
    except Exception as e:
        return f"Error: {str(e)}"

# Function to compile and run Java code
def run_java(code):
    try:
        file_path = os.path.join(CODE_DIR, "Main.java")
        with open(file_path, 'w') as f:
            f.write(code)

        # Compile Java code
        compile_result = subprocess.run(
            ['javac', file_path],
            capture_output=True,
            text=True,
            timeout=10
        )
        if compile_result.stderr:
            return compile_result.stderr

        # Run Java program
        run_result = subprocess.run(
            ['java', '-cp', CODE_DIR, 'Main'],
            capture_output=True,
            text=True,
            timeout=10
        )
        return run_result.stdout or run_result.stderr
    except subprocess.TimeoutExpired:
        return "Error: Code execution timed out."
    except Exception as e:
        return f"Error: {str(e)}"

# Function to compile and run C code
def run_c(code):
    try:
        file_path = os.path.join(CODE_DIR, "main.c")
        output_path = os.path.join(CODE_DIR, "main")
        with open(file_path, 'w') as f:
            f.write(code)

        # Compile C code
        compile_result = subprocess.run(
            ['gcc', '-o', output_path, file_path],
            capture_output=True,
            text=True,
            timeout=10
        )
        if compile_result.stderr:
            return compile_result.stderr

        # Execute C program
        run_result = subprocess.run(
            [output_path],
            capture_output=True,
            text=True,
            timeout=10
        )
        return run_result.stdout or run_result.stderr
    except subprocess.TimeoutExpired:
        return "Error: Code execution timed out."
    except Exception as e:
        return f"Error: {str(e)}"

# Function to compile and run C++ code
def run_cpp(code):
    try:
        file_path = os.path.join(CODE_DIR, "main.cpp")
        output_path = os.path.join(CODE_DIR, "main")
        with open(file_path, 'w') as f:
            f.write(code)

        # Compile C++ code
        compile_result = subprocess.run(
            ['g++', '-o', output_path, file_path],
            capture_output=True,
            text=True,
            timeout=10
        )
        if compile_result.stderr:
            return compile_result.stderr

        # Execute C++ program
        run_result = subprocess.run(
            [output_path],
            capture_output=True,
            text=True,
            timeout=10
        )
        return run_result.stdout or run_result.stderr
    except subprocess.TimeoutExpired:
        return "Error: Code execution timed out."
    except Exception as e:
        return f"Error: {str(e)}"

# Function to run JavaScript code
def run_javascript(code):
    try:
        file_path = os.path.join(CODE_DIR, "script.js")
        with open(file_path, 'w') as f:
            f.write(code)

        # Run JavaScript code using Node.js
        result = subprocess.run(
            ['node', file_path],
            capture_output=True,
            text=True,
            timeout=10
        )
        return result.stdout or result.stderr
    except subprocess.TimeoutExpired:
        return "Error: Code execution timed out."
    except Exception as e:
        return f"Error: {str(e)}"

@app.route('/run-code', methods=['POST'])
def run_code():
    data = request.get_json()
    language = data.get('language')
    code = data.get('code')

    if not language or not code:
        return jsonify({'output': 'Error: Language and code are required.'}), 400

    try:
        if language == 'python':
            output = run_python(code)
        elif language == 'java':
            output = run_java(code)
        elif language == 'c':
            output = run_c(code)
        elif language == 'cpp':
            output = run_cpp(code)
        elif language == 'javascript':
            output = run_javascript(code)
        else:
            output = "Error: Unsupported language."
        return jsonify({'output': output})
    except Exception as e:
        return jsonify({'output': f"Error: {str(e)}"}), 500
    finally:
        cleanup()

if __name__ == '__main__':
    app.run(debug=True)

File: compilerserver.js
const express = require('express');
const cors = require('cors');
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');
const os = require('os');

const app = express();
const CODE_DIR = path.join(__dirname, 'code_exec');

// Middleware
app.use(express.json());
app.use(cors());

// Ensure temporary directory exists
if (!fs.existsSync(CODE_DIR)) {
    fs.mkdirSync(CODE_DIR, { recursive: true });
}

// Cleanup function
const cleanup = () => {
    fs.rmSync(CODE_DIR, { recursive: true, force: true });
    fs.mkdirSync(CODE_DIR, { recursive: true });
};

// Function to run Python code
const runPython = (code) => {
    return new Promise((resolve) => {
        const filePath = path.join(CODE_DIR, "script.py");
        fs.writeFileSync(filePath, code);

        exec(`python3 "${filePath}"`, { timeout: 5000 }, (error, stdout, stderr) => {
            resolve(stdout.trim() || stderr.trim() || error?.message);
        });
    });
};


// Function to compile & run Java code
const runJava = (code) => {
    return new Promise((resolve) => {
        const filePath = path.join(CODE_DIR, 'Main.java');
        fs.writeFileSync(filePath, code);   

        exec(`javac ${filePath}`, { timeout: 10000 }, (error, stdout, stderr) => {
            if (stderr) return resolve(stderr);

            exec(`java -cp ${CODE_DIR} Main`, { timeout: 10000 }, (error, stdout, stderr) => {
                resolve(stdout || stderr || error?.message);
            });
        });
    });
};

// Function to compile & run C code
const runC = (code) => {
    return new Promise((resolve) => {
        const filePath = path.join(CODE_DIR, 'main.c');
        const outputPath = path.join(CODE_DIR, 'main');
        fs.writeFileSync(filePath, code);

        exec(`gcc -o ${outputPath} ${filePath}`, { timeout: 10000 }, (error, stdout, stderr) => {
            if (stderr) return resolve(stderr);

            exec(outputPath, { timeout: 10000 }, (error, stdout, s3tderr) => {
                resolve(stdout || stderr || error?.message);
            });
        });
    });
};

// Function to compile & run C++ code
const runCpp = (code) => {
    return new Promise((resolve) => {
        const filePath = path.join(CODE_DIR, 'main.cpp');
        const outputPath = path.join(CODE_DIR, 'main');
        fs.writeFileSync(filePath, code);

        exec(`g++ -o ${outputPath} ${filePath}`, { timeout: 10000 }, (error, stdout, stderr) => {
            if (stderr) return resolve(stderr);

            exec(outputPath, { timeout: 10000 }, (error, stdout, stderr) => {
                resolve(stdout || stderr || error?.message);
            });
        });
    });
};

// Function to run JavaScript code using Node.js
const runJavaScript = (code) => {
    return new Promise((resolve) => {
        const filePath = path.join(CODE_DIR, "script.js");
        fs.writeFileSync(filePath, code);

        exec(`node "${filePath}"`, { timeout: 10000 }, (error, stdout, stderr) => {
            resolve(stdout.trim() || stderr.trim() || error?.message);
        });
    });
};

// API Endpoint to run code
app.post('/run-code', async (req, res) => {
    const { language, code } = req.body;
    if (!language || !code) {
        return res.status(400).json({ output: 'Error: Language and code are required.' });
    }

    try {
        let output;
        switch (language) {
            case 'python':
                output = await runPython(code);
                break;
            case 'java':
                output = await runJava(code);
                break;
            case 'c':
                output = await runC(code);
                break;
            case 'cpp':
                output = await runCpp(code);
                break;
            case 'javascript':
                output = await runJavaScript(code);
                break;
            default:
                output = 'Error: Unsupported language.';
        }

        return res.json({ output });
    } catch (error) {
        return res.status(500).json({ output: `Error: ${error.message}` });
    } finally {
        cleanup();
    }
});

// Function to get the device's IP address
const getDeviceIp = () => {
    const interfaces = os.networkInterfaces();
    for (const interfaceName of Object.keys(interfaces)) {
        for (const iface of interfaces[interfaceName]) {
            if (iface.family === 'IPv4' && !iface.internal) {
                return iface.address;
            }
        }
    }
    return '0.0.0.0';
};

// Start the server
const PORT = process.env.PORT || 3000;
const IP = getDeviceIp();
app.listen(PORT, '0.0.0.0', () => {
    console.log(`Server running on http://${IP}:${PORT}`);
});